File: androidApp/src/main/java/com/wonddak/loacell/android/ui/setting/LoginInfoView.kt
package com.wonddak.loacell.android.ui.setting

import android.app.Activity
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Divider
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.wonddak.loacell.SharedRes
import com.wonddak.loacell.android.ui.common.MyIconButton
import com.wonddak.loacell.android.viewModel.LoaCellViewModel
import com.wonddak.loacell.auth.delete
import com.wonddak.loacell.model.DialogStatus
import kotlinx.coroutines.launch

@Composable
fun LoginInfoView(
    loaCellViewModel: LoaCellViewModel
) {
    val loginHelper = loaCellViewModel.loginHelper
    val user by loaCellViewModel.user.collectAsState(null)
    val roomLists by loaCellViewModel.roomList.collectAsState()
    val roomList = roomLists.filter { it.owner == user?.uid }

    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    user?.let { userInfo ->
        Column(
            modifier = Modifier.fillMaxWidth()
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Column {
                    Text(
                        text = (userInfo.displayName ?: "").ifEmpty{ "이름 없음" },
                        fontWeight = FontWeight.Bold
                    )
                    Text(
                        text = userInfo.uid
                    )
                }
                MyIconButton(imageResource = SharedRes.images.change_person) {
                    loaCellViewModel.showDialog(DialogStatus.SETTING_EDIT_NAME)
                }
            }
            Divider()

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(5.dp)
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(5.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    val buttonWeight = Modifier.weight(1f)

                    OutlinedButton(
                        modifier = buttonWeight,
                        onClick = {
                            loaCellViewModel.outOrSignOut()
                        }
                    ) {
                        Text(text = if (userInfo.isAnonymous) "나가기" else "로그아웃")
                    }
                    if (!userInfo.isAnonymous) {
                        Spacer(modifier = Modifier.width(15.dp))
                        OutlinedButton(
                            modifier = buttonWeight,
                            onClick = {
                                loginHelper.delete()
                            },
                            enabled = roomList.isEmpty()
                        ) {
                            Text(text = "탈퇴")
                        }
                    }
                }
                if (!userInfo.isAnonymous && roomList.isNotEmpty()) {
                    Text(text = "소유자인 방의 정보를 모두 삭제해 주세요")
                }
                if (userInfo.isAnonymous) {
                    OutlinedButton(
                        onClick = {
                            scope.launch {
                                loginHelper.linkToGoogle(
                                    context as Activity,
                                    failAction = { msg ->
                                        loaCellViewModel.showSnackBar(msg)
                                    }
                                ) {
                                    loaCellViewModel.closeSetting()
                                }
                            }
                        }
                    ) {
                        Text(text = "Google로 연동")
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/ui/setting/LoginInfoView.swift
//
//  LoginInfoView.swift
//  iosApp
//
//  Created by WonHee Jung on 2023/11/17.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared
import FirebaseAuth

struct LoginInfoView: View {
    @EnvironmentObject var viewModel: LoaCellViewModel
    
    var user : FBUser?  {
        viewModel.user
    }
    var loginHelper : LoginHelper {
        self.viewModel.loginHelper
    }
    var myRoomList : [RoomInfo] {
        viewModel.roomList.filter { info in
            info.owner == user?.uid
        }
    }
    
    @Environment(\.window) var window: UIWindow?
    
    @State private var showDeleteError = false
    @State private var appleLinkCoordinator: AppleLinkCoordinator?
    @State private var appleRevokeCoordinator :AppleTokenRevokeCoordinator?
    @State private var isAppleProvider :Bool = false
    
    private let iconSize : CGFloat = 36
    var body: some View {
        VStack {
            if let userInfo = user {
                HStack {
                    VStack(alignment: .leading) {
                        Text((userInfo.displayName ?? "").ifEmpty{ "이름 없음"})
                            .fontWeight(.bold)
                        Text(userInfo.uid)
                    }
                    Spacer()
                    IconButton(resource: \.change_person) {
                        viewModel.showDialog(dialogStatus: DialogStatus.settingEditName)
                    }
                }
                Divider()
                
                VStack {
                    HStack {
                        RoundCornerButton(text: userInfo.isAnonymous ? "나가기" : "로그아웃") {
                            viewModel.outOrSignOut()
                        }
                        if !userInfo.isAnonymous {
                            Spacer()
                                .frame(width: 15)
                            RoundCornerButton(text: "탈퇴") {
                                deleteAccount()
                            }
                        }
                    }
                    if showDeleteError {
                        Text("소유자인 방의 정보를 모두 삭제해 주세요")
                            .onAppear {
                                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 1.5) {
                                    showDeleteError = false
                                }
                            }
                    }
                    if userInfo.isAnonymous {
                        HStack {
                            Text("연동하기")
                            Spacer()
                                .frame(width: 10)
                            Button {
                                linkToGoogle()
                            } label: {
                                Image(resource: \.btn_google)
                                    .resizable()
                                    .frame(width: iconSize, height: iconSize)
                                
                            }
                            Button {
                                linkToApple()
                            } label: {
                                Image(resource: \.logo_apple)
                                    .resizable()
                                    .frame(width: iconSize, height: iconSize)
                                
                            }
                            Spacer()
                        }
                      

                    }
                }
                .frame(maxWidth: .infinity)
                .padding(5)
            }
        }
        .onAppear {
            Auth.auth().currentUser?.providerData.forEach({ info in
                print(info.providerID)
                if (info.providerID == "apple.com") {
                    isAppleProvider = true
                }
            })
        }
    }
    
    private func linkToGoogle() {
        loginHelper.requestAnonymousToGoogleLogin(
            failAction: { msg in
                viewModel.showSnackBar(msg: msg)
            },successAction: {
                
            }
        )
    }
    
    private func linkToApple() {
        appleLinkCoordinator = AppleLinkCoordinator(
            window: window,
            linkFailAction: { msg in
                viewModel.showSnackBar(msg: msg)
            },
            linkSuccessAction: {
                viewModel.closeSetting()
            })
        appleLinkCoordinator?.startLogin()
    }
    
    private func deleteAccount() {
        if !myRoomList.isEmpty {
            showDeleteError = true
            return
        }
        if !isAppleProvider {
            loginHelper.delete()
            return
        }
        appleRevokeCoordinator = AppleTokenRevokeCoordinator(window: window, failAction: { msg in
            viewModel.showSnackBar(msg: msg)
        }, revokeSuccessAction: {
            print("revoke Success")
            loginHelper.delete()
        })
        appleRevokeCoordinator?.startLogin()
    }
}

